### PR TITLE
Prevent .part upload via dav

### PR DESCRIPTION
**Pull request by IljaN**
## Description
Prevent creating, moving, copying to paths with extension .part
## Related Issue
#7496 

## How Has This Been Tested?
Manually, Integration-Tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by mmattel** [comment]
@IljaN 
one question, what would happen if somone creates a folder called ``.part`` ?

**Comment by IljaN** [comment]
@mmattel Good catch! It is not possible via webdav: mkcol .part Fails with 405 Method not allowed, not sure if it is the correct code

If you create the directory directly in your storage with mkdir .part it is ignored by owncloud since we don`t support dotfiles afaik

**Comment by PVince81** [comment]
@iljan we should also block "$name.part" folders. Ignoring is not good. If you had the impression that OC ignores it, did you check the FS ? Maybe it created the folder still but it wasn't visible in the UI ?

Better block those as well.

**Comment by mmattel** [comment]
@IljaN maybe you could use from https://github.com/owncloud/core/pull/19235 
``lib/private/Files/Filesystem.php`` function ``isForbiddenFileOrDir`` and add ``.part`` into the query array

**Comment by IljaN** [comment]
@PVince81 .part directory creation is already prevented by this change:
```
dav:/remote.php/dav/files/admin/> mkcol path.with.ext.part
Creating `path.with.ext.part': failed:
400 Bad Request
```
@mmattel I extended verfiyPath method with a extension check, this seems to catch all cases with the benefit that it is used in all codepaths (Single point of truth), what benefits does extending isForbiddenFileOr dir give us?

**Comment by PVince81** [comment]
@IljaN prove it with an integration test 😉 

Is 400 bad request the correct code, same as for files ? Just asking, I thought we returned 403 for those but don't remember exactly.

**Comment by mmattel** [comment]
@IljaN nothing special, just remembered that we excluded stuff some time ago and thought it is maybe usable for your PR, never mind

**Comment by PVince81** [comment]
rebase conflicts

**Comment by IljaN** [comment]
@PVince81 let me fix them, I am also in the process adding a missing test

**Comment by IljaN** [comment]
@PVince81 Ready

File: lib/private/Files/View.php [hunk 1]
```diff
@@ -1799,6 +1799,13 @@ public function verifyPath($path, $fileName) {
 			throw new InvalidPathException($l10n->t('Dot files are not allowed'));
 		}
 
+
+		if (pathinfo($fileName, PATHINFO_EXTENSION) == 'part') {
```
Comment (PVince81) [inline lib/private/Files/View.php]: did we talk about https://github.com/owncloud/core/pull/27149/files#diff-6cc3350c00dd13da023086587cce8bfdR64 ?

**Comment by IljaN** [comment]
Ready for next review round

File: lib/private/Files/View.php [hunk 2]
```diff
@@ -1799,6 +1799,13 @@ public function verifyPath($path, $fileName) {
 			throw new InvalidPathException($l10n->t('Dot files are not allowed'));
 		}
 
+
+		if (preg_match('/' . FileInfo::BLACKLIST_FILES_REGEX . '/', $fileName) !== 0) {
+			throw new InvalidPathException(
+				'Can`t upload files with extension .part because this extension is used internally by owncloud.'
```
Comment (mmattel) [inline lib/private/Files/View.php]: wouldnt it be good to get the filename/component in question from the constant instead of hardcoding it?

Comment (PVince81) [reply lib/private/Files/View.php]: @IljaN see above comment

Also we cannot use the name "ownCloud" in error messages due to theming, please remove it

Comment (IljaN) [reply lib/private/Files/View.php]: Not sure if I understand I am getting it from FileInfo::BLACKLIST_FILES_REGEX

Comment (mmattel) [reply lib/private/Files/View.php]: To be precise, it is not the query but the text.

Here you state hardcoded ``.part`` 
But the regex says ``const BLACKLIST_FILES_REGEX = '\.(part|filepart)$';``
One name is missing.
What happens when the regex is extended with more names?
Extract the names and print them to be futureproof.

**Comment by IljaN** [comment]
Ready for next round of review

**Review by PVince81 — APPROVED** [review]
:+1: looks good to me, thanks a lot!

**Comment by PVince81** [comment]
@IljaN please backport to stable10

**Comment by PVince81** [comment]
stable10: https://github.com/owncloud/core/pull/29432

**Pull request merged**